File: linda-presentation/src/main/java/it/unibo/coordination/linda/presentation/DynamicDeserializer.java
package it.unibo.coordination.linda.presentation;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.Reader;
import java.util.List;

public abstract class DynamicDeserializer<T> extends SimpleDeserializer<T> {

    public DynamicDeserializer(Class<T> clazz, MIMETypes mimeType, ObjectMapper mapper) {
        super(clazz, mimeType, mapper);
    }

    @Override
    public abstract T fromDynamicObject(Object dynamicObject);

    @Override
    public T read(Reader reader) {
        return fromDynamicObject(readImpl(reader, Object.class));
    }

    @Override
    public List<T> readList(Reader reader) {
        return listFromDynamicObject(readImpl(reader, Object.class));
    }
}

File: presentation/src/main/kotlin/it/unibo/presentation/DynamicDeserializer.kt
package it.unibo.presentation

import com.fasterxml.jackson.databind.ObjectMapper
import java.io.Reader

abstract class DynamicDeserializer<T>(clazz: Class<T>, mimeType: MIMETypes, mapper: ObjectMapper) : SimpleDeserializer<T>(clazz, mimeType, mapper) {

    abstract override fun fromDynamicObject(dynamicObject: Any): T

    override fun read(reader: Reader): T {
        return fromDynamicObject(readImpl(reader, Any::class.java))
    }

    override fun readList(reader: Reader): List<T> {
        return listFromDynamicObject(readImpl(reader, Any::class.java))
    }
}